improve overloads support, attempt 2 (#83)
mmkal/expect-type: README.md

// Instead of the above, try something like this:
expectTypeOf(B).instance.toEqualTypeOf<{b: string; foo: () => void}>()
```

> Overloads limitation: overloaded functions which include an overload looking like `(...args: unknown[]) => unknown` will excluded `unknown[]` from `.parameters` and exclude `unknown` from `.returns`. This overload just allows any input and returns unknown output anyway though.:

Reviewer: ```suggestion
Overloads limitation: Overloaded functions that include an overload resembling `(...args: unknown[]) => unknown` will exclude `unknown[]` from `.parameters` and `unknown` from `.returns`. This overload, however, allows any input and returns an unknown output anyway:
```

Author: re-jigged this a little since you commented but done.

- built into existing tooling. No extra build step, cli tool, IDE extension, or lint plugin is needed. Just import the function and start writing tests. Failures will be at compile time - they'll appear in your IDE and when you run `tsc`.
- small implementation with no dependencies. [Take a look!](./src/index.ts) (tsd, for comparison, is [2.6MB](https://bundlephobia.com/result?p=tsd@0.13.1) because it ships a patched version of typescript).

## TypeScript backwards-compatibility

> There is a CI job `test-types` which checks that the tests still pass with certain older typescript versions. To check supported typescript versions, [look at the job definition](./.github/workflows/ci.yml).

Reviewer: ```suggestion
There is a CI job called `test-types` that checks whether the tests still pass with certain older TypeScript versions. To check the supported TypeScript versions, [refer to the job definition](./.github/workflows/ci.yml).
```

If you're adding a feature, you should write a self-contained usage example in the form of a test, in [test/usage.test.ts](./test/usage.test.ts). This file is used to populate the bulk of this readme using [eslint-plugin-codegen](https://npmjs.com/package/eslint-plugin-codegen), and to generate an ["errors" test file](./test/errors.test.ts), which captures the error messages that are emitted for failing assertions by the typescript compiler. So, the test name should be written as a human-readable sentence explaining the usage example. Have a look at the existing tests for an idea of the style.

After adding the tests, run `npm run lint -- --fix` to update the readme, and `npm test -- --updateSnapshot` to update the errors test. The generated documentation and tests should be pushed to the same branch as the source code, and submitted as a pull request. CI will test that the docs and tests are up to date if you forget to run these commands.

### Documentation of limitations through tests

> Limitations of the library are documented via tests in usage.test.ts. This means that if a future TypeScript version (or library version) fixes the limitation, the test will start failing, so it will be automatically removed from the docs once it no longer applies.

Reviewer: ```suggestion
Limitations of the library are documented through tests in `usage.test.ts`. This means that if a future TypeScript version (or library version) fixes the limitation, the test will start failing, and it will be automatically removed from the documentation once it no longer applies.
```
